src/main/objects/Parser.js

```diff
@@ -27,24 +30,125 @@ function parseChord(string) {
     var arguments_json = JSON.parse(arguments);
     arguments_json["functionName"] = chord_type[0];
     arguments_json["mode"] = mode;
-    var ret = new HarmonicFunction.HarmonicFunction2(arguments_json);
+    return new HarmonicFunction.HarmonicFunction2(arguments_json)
+}
 
-    return ret
+function getKeyFromPitchBasenoteAndModeOrThrowError(pitch, basenote, mode) {
+    var mapKey = pitch.toString() + "," + basenote.toString() + "," + mode
+    var key = Consts.keyFromPitchBasenoteAndMode[mapKey]
+    if (key === undefined) {
+        throw new Errors.ProbablyUnexpectedError("Could not find key for given pitch, basenote and mode",
+            "Pitch: " + pitch + " Basenote: " + basenote + " Mode: " + mode)
+    } else {
+        return key
+    }
+}
+
+function calculateKey(key, nextChordAfterDeflection) {
+
+    var keyToUse = key
+    if (nextChordAfterDeflection.key !== undefined) {
+        keyToUse = nextChordAfterDeflection.key
+    }
+
+    var pitchesToUse = Utils.contains(Consts.possible_keys_major, keyToUse) ?
+        [0, 2, 4, 5, 7, 9, 11] : [0, 2, 3, 5, 7, 8, 10]
```

Comment:
Tu może lepiej użyć Scale.MajorScale itp?

Reply:
zapomniałem, że miałem to zrobić :P już zmieniam